### PR TITLE
feat: integrate authenticated calls smart contract functionality into protocol

### PR DESCRIPTION
**Pull request by skosito**
# Description

Integrates this https://github.com/zeta-chain/protocol-contracts/issues/344 into protocol, this PR should be merged after smart contract changes are merged, but this is ready for first pass.

Also e2e tests to call zevm->evm, using call and withdrawAndCall and authenticated sender, with checking if sender is properly forwarded, so destination contract can use it as authentication.

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Summary by CodeRabbit

- **New Features**
	- Introduced new test cases for Ethereum and ZEVM authenticated calls, enhancing the testing framework.
	- Added new functions in the `TestDAppV2` contract for improved message handling and sender management.
	- New `TestGatewayZEVMCaller` contract facilitates interactions with a gateway on the ZEVM.

- **Bug Fixes**
	- Updated function signatures to improve parameter handling and streamline event processing.

- **Documentation**
	- Updated comments and descriptions to clarify new functionalities and parameters across various files.

- **Chores**
	- Updated dependency versions in `go.mod` for improved stability and performance.

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -11,6 +11,7 @@
 * [2861](https://github.com/zeta-chain/node/pull/2861) - emit events from staking precompile
 * [2870](https://github.com/zeta-chain/node/pull/2870) - support for multiple Bitcoin chains in the zetaclient
 * [2883](https://github.com/zeta-chain/node/pull/2883) - add chain static information for btc signet testnet
+* [2904](https://github.com/zeta-chain/node/pull/2904) - authenticated zevm -> evm call
```
Comment (lumtis) [inline changelog.md:14]: "integrate authenticated calls smart contract functionality into protocol"

File: cmd/zetae2e/local/v2.go [hunk 2]
```diff
@@ -20,7 +20,11 @@ func startV2Tests(eg *errgroup.Group, conf config.Config, deployerRunner *runner
 		e2etests.TestV2ETHDepositAndCallName,
 		e2etests.TestV2ETHWithdrawName,
 		e2etests.TestV2ETHWithdrawAndCallName,
+		e2etests.TestV2ETHWithdrawAndAuthenticatedCallName,
+		e2etests.TestV2ETHWithdrawAndAuthenticatedCallThroughContractName,
```
Comment (lumtis) [inline cmd/zetae2e/local/v2.go:24]: Might be refactored in a further PR but I think we set these as the standard withdraw test and rename the current withdraw tests into `WithdrawArbitraryCall`

File: e2e/runner/v2_zevm.go [hunk 3]
```diff
@@ -50,6 +52,56 @@ func (r *E2ERunner) V2ETHWithdrawAndCall(
 	return tx
 }
 
+// V2ETHWithdrawAndCall calls WithdrawAndCall of Gateway with gas token on ZEVM using authenticated call
+func (r *E2ERunner) V2ETHWithdrawAndAuthenticatedCall(
```
Comment (lumtis) [inline e2e/runner/v2_zevm.go:56]: Same comment as above here, rename the old function "arbitraryCall"

File: pkg/contracts/testgatewayzevmcaller/TestGatewayZEVMCaller.sol [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+interface IZRC20 {
+    function approve(address spender, uint256 amount) external returns (bool);
+}
+
+interface WZETA {
+    function deposit() external payable;
+    function approve(address guy, uint256 wad) external returns (bool);
+}
+
+contract TestGatewayZEVMCaller {
```
Comment (lumtis) [inline pkg/contracts/testgatewayzevmcaller/TestGatewayZEVMCaller.sol:57]: I think we can simplify and remove the test prefix

File: proto/zetachain/zetacore/crosschain/tx.proto [hunk 5]
```diff
@@ -175,6 +175,8 @@ message MsgVoteInbound {
 
   // revert options provided by the sender
   RevertOptions revert_options = 17 [ (gogoproto.nullable) = false ];
+
+  bool is_arbitrary_call = 18;
```
Comment (lumtis) [inline proto/zetachain/zetacore/crosschain/tx.proto:179]: Can we maybe instead have a `call_options` param? And add comment for the field

File: proto/zetachain/zetacore/crosschain/cross_chain_tx.proto [hunk 6]
```diff
@@ -79,6 +79,7 @@ message OutboundParams {
   uint64 effective_gas_limit = 22;
   string tss_pubkey = 11;
   TxFinalizationStatus tx_finalization_status = 12;
+  bool is_arbitrary_call = 24;
```
Comment (lumtis) [inline proto/zetachain/zetacore/crosschain/cross_chain_tx.proto:82]: Same as above

File: x/crosschain/client/cli/tx_vote_inbound.go [hunk 7]
```diff
@@ -83,6 +83,7 @@ func CmdVoteInbound() *cobra.Command {
 				argsAsset,
 				uint(argsEventIndex),
 				protocolContractVersion,
+				true, // TODO: do we need to provide this as arg?
```
Comment (lumtis) [inline x/crosschain/client/cli/tx_vote_inbound.go:86]: Yes we should

File: x/crosschain/keeper/evm_hooks.go [hunk 8]
```diff
@@ -206,6 +206,7 @@ func (k Keeper) ProcessZRC20WithdrawalEvent(
 		foreignCoin.Asset,
 		event.Raw.Index,
 		types.ProtocolContractVersion_V1,
+		true, // not relevant for v1
```
Comment (lumtis) [inline x/crosschain/keeper/evm_hooks.go:209]: Shouldn't we rather set to false even if not used? So zero value is used

File: x/crosschain/keeper/v2_zevm_inbound.go [hunk 9]
```diff
@@ -183,7 +181,7 @@ func (k Keeper) newWithdrawalInbound(
 		return nil, errors.Wrapf(err, "cannot encode address %v", event.Receiver)
 	}
 
-	gasLimit := event.GasLimit.Uint64()
+	gasLimit := event.CallOptions.GasLimit.Uint64()
```
Comment (lumtis) [inline x/crosschain/keeper/v2_zevm_inbound.go:184]: There will be a delay between:
- processing upgrade that will use this new logic
- upgrading the gateway contract to emit the event with callOptions

In practice we should upgrade immediately after the chain upgrade, and pause cctx in between but we should still define the logic handling the event when the contract is not upgrade yet.

Do we have some mechanism to for example reject the tx when the event is not up to date (so not new cctx before contract upgraded)

File: x/crosschain/keeper/v2_zevm_inbound.go [hunk 10]
```diff
@@ -197,7 +196,7 @@ func (k Keeper) newWithdrawalInbound(
 
 	return types.NewMsgVoteInbound(
 		"",
-		from.Hex(),
+		event.Sender.Hex(),
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/crosschain/keeper/v2_zevm_inbound.go:198]: Looks to make more sense using the event

File: zetaclient/chains/evm/observer/v2_inbound.go [hunk 11]
```diff
@@ -325,6 +326,7 @@ func (ob *Observer) newCallInboundVote(event *gatewayevm.GatewayEVMCalled) types
 		"",
 		event.Raw.Index,
 		types.ProtocolContractVersion_V2,
+		false, // currently not relevant since calls are not arbitrary
```
Comment (kingpinXD) [inline zetaclient/chains/evm/observer/v2_inbound.go:329]: Can we define this as a constant ?

File: x/crosschain/keeper/v2_zevm_inbound.go [hunk 12]
```diff
@@ -68,12 +67,12 @@ func (k Keeper) ProcessZEVMInboundV2(
...
 		// create inbound object depending on the event type
 		if withdrawalEvent != nil {
-			inbound, err = k.newWithdrawalInbound(ctx, from, txOrigin, foreignCoin, withdrawalEvent)
+			inbound, err = k.newWithdrawalInbound(ctx, txOrigin, foreignCoin, withdrawalEvent)
 			if err != nil {
 				return err
 			}
 		} else {
-			inbound, err = k.newCallInbound(ctx, from, txOrigin, foreignCoin, gatewayEvent)
+			inbound, err = k.newCallInbound(ctx, txOrigin, foreignCoin, gatewayEvent)
```
Comment (kingpinXD) [inline x/crosschain/keeper/v2_zevm_inbound.go:75]: Not required for this PR but consider refactoring to extract the common logic out of Withdraw and Call Inbound creation into a separate function .

File: e2e/e2etests/test_v2_eth_withdraw_and_authenticated_call.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+	crosschaintypes "github.com/zeta-chain/node/x/crosschain/types"
+)
+
+const payloadMessageAuthenticatedWithdrawETH = "this is a test ETH withdraw and authenticated call payload"
+
+func TestV2ETHWithdrawAndAuthenticatedCall(r *runner.E2ERunner, args []string) {
+	require.Len(r, args, 1)
+
+	previousGasLimit := r.ZEVMAuth.GasLimit
+	r.ZEVMAuth.GasLimit = 10000000
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply e2e/e2etests/test_v2_eth_withdraw_and_authenticated_call.go:21]: I think this is a good idea

File: e2e/runner/v2_zevm.go [hunk 14]
```diff
@@ -103,7 +155,7 @@ func (r *E2ERunner) V2ZEVMToEMVCall(
 	payload []byte,
 	revertOptions gatewayzevm.RevertOptions,
 ) *ethtypes.Transaction {
-	tx, err := r.GatewayZEVM.Call(
+	tx, err := r.GatewayZEVM.Call0(
```
Comment (lumtis) [inline e2e/runner/v2_zevm.go:158]: Did the function name to use changed randomly after changing the commit used for protocol contracts?

I had this, was a bit annoying, wondering if we could make the overloaded function Go binding more determistic

**Review by lumtis — APPROVED** [review]
LGTM

**Comment by lumtis** [comment]
Upgrade test likely to fail because of https://github.com/zeta-chain/node/issues/2921

I think it should be trivial I will also have a look

**Pull request merged**